client/src/js/components/bhCardList.js

```diff
@@ -0,0 +1,100 @@
+angular.module('bhima.components')
+  .component('bhCardList', {
+    templateUrl : 'modules/templates/bhCardList.tmpl.html',
+    controller : bhCardList,
+    bindings : {
+      data : '<',
+      template : '@',
+      name : '@', // name attribute to be sorted alphabetically and filtered
+      age : '@?', // date attribute to be sorted by date
+      size : '@?', // size attribute to be sorted largest to smallest
+      id : '@?', // optional data override
+    }
+    ,
+  });
+
+/**
+ * @description
+ * Component for displaying cards of information, providing sorting and filtering
+ * functionality.
+ *
+ * Requires:
+ * `data` : the array to iterate over
+ * `name` : the entity display name that should be sorted and filtered
+ * `template` : path to an HTML template that should be displayed for each element
+ *
+ * Optional:
+ * `age` : the entity date field key, this will allow it to be sorted by age
+ * `size` : the entity size field key, this will allow it to be sorted by largest
+ * `id` : provide a custom entity identifier override
+ */
+function bhCardList() {
+  const $ctrl = this;
+
+  // name, age and size options will be added to this _if_ they are defined
+  // through bindings
+  $ctrl.orderOptions = { active : false, available : [] };
+  $ctrl.filterOptions = { active : false };
+
+  $ctrl.$onInit = function onInit() {
+    $ctrl.identifier = $ctrl.id;
+
+    assignAvailableOrders($ctrl.name, $ctrl.age, $ctrl.size);
+  };
+
+  $ctrl.$onChanges = function onChanges(changes) {
+    if (changes.data && angular.isDefined(changes.data.currentValue)) {
+      if (!$ctrl.identifier) {
+        // no identifier override found
+        $ctrl.identifier = defaultIdentifier(changes.data.currentValue);
+      }
+    }
+  };
+
+  $ctrl.toggleFilter = function toggleFilter() {
+    $ctrl.filterOptions.active = !$ctrl.filterOptions.active; $ctrl.filterOptions.value = {};
+  };
+
+  $ctrl.setOrder = function setOrder(order) {
+    $ctrl.orderOptions.active = angular.isDefined(order);
+    $ctrl.orderOptions.current = order || $ctrl.orderOptions.defaultOrder;
+  };
+
+  // attempt to pick an identifier based on common ids
+  // this is overriden by binding a specific `id`
+  // supported default identifiers are 'id' and 'uuid'
+  function defaultIdentifier(data) {
+    const sampleData = data[0] || {};
+    return angular.isDefined(sampleData.uuid) ? 'uuid' : 'id';
+  }
+
+  function assignAvailableOrders(name, age, size) {
+    // name binding is required - assume this is valid
+    $ctrl.orderOptions.available.push(
+      { attribute : name, key : 'TABLE.COLUMNS.SORTING.NAME_ASC', reverse : false }
+    );
+    $ctrl.orderOptions.available.push(
+      { attribute : name, key : 'TABLE.COLUMNS.SORTING.NAME_DSC', reverse : true }
+    );
+
+    // parse optional bindings
+    if (angular.isDefined(age)) {
+      $ctrl.orderOptions.available.push(
```

Comment:
Just FYI, you can pass multiple parameters into `Array.prototype.push`, so you could write:
```js
$ctrl.orderOptions.available.push(
  { attribute : age, key : 'TABLE.COLUMNS.SORTING.CREATED_ASC', reverse : false },
  { attribute : age, key : 'TABLE.COLUMNS.SORTING.CREATED_DSC', reverse : true }
);
```
to achieve the same result.